File: modules/zig/src/main/java/com/falsepattern/zigbrains/zig/util/ZigStringUtil.java
package com.falsepattern.zigbrains.zig.util;

import com.intellij.openapi.util.Pair;
import com.intellij.openapi.util.TextRange;
import it.unimi.dsi.fastutil.ints.Int2IntMap;
import it.unimi.dsi.fastutil.ints.Int2IntOpenHashMap;
import lombok.experimental.UtilityClass;
import lombok.val;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

public class ZigStringUtil {
    public static String escape(String input) {
        return input.codePoints().mapToObj(point -> switch (point) {
            case '\n' -> "\\n";
            case '\r' -> "\\r";
            case '\t' -> "\\t";
            case '\\' -> "\\\\";
            case '"' -> "\\\"";
            case '\'', ' ', '!' -> Character.toString(point);
            default -> {
                if (point >= '#' && point <= '&' ||
                    point >= '(' && point <= '[' ||
                    point >= ']' && point <= '~') {
                    yield Character.toString(point);
                } else {
                    yield "\\u{" + Integer.toHexString(point).toLowerCase() + "}";
                }
            }
        }).collect(Collectors.joining(""));
    }

    public static List<Pair<TextRange, String>> getDecodeReplacements(@NotNull CharSequence input, boolean isMultiline) {
        if (isMultiline) {
            return List.of();
        }

        val result = new ArrayList<Pair<TextRange, String>>();
        for (int i = 0; i + 1 < input.length(); i++) {
            if (input.charAt(i) == '\\') {
                val length = Escaper.findEscapementLength(input, i);
                val charCode = Escaper.toUnicodeChar(input, i, length);
                val range = TextRange.create(i, Math.min(i + length + 1, input.length()));
                result.add(Pair.create(range, Character.toString(charCode)));
                i += range.getLength() - 1;
            }
        }
        return result;
    }

    public static String unescape(@NotNull CharSequence input, boolean isMultiline) {
        return isMultiline ? input.toString() : processReplacements(input, getDecodeReplacements(input, false));
    }

    private static @NotNull String processReplacements(@NotNull CharSequence input,
                                                       @NotNull List<? extends Pair<TextRange, String>> replacements) throws IndexOutOfBoundsException {
        StringBuilder result = new StringBuilder();
        int currentOffset = 0;
        for (val replacement: replacements) {
            result.append(input.subSequence(currentOffset, replacement.getFirst().getStartOffset()));
            result.append(replacement.getSecond());
            currentOffset = replacement.getFirst().getEndOffset();
        }
        result.append(input.subSequence(currentOffset, input.length()));
        return result.toString();
    }

    private static final Pattern NL_MATCHER = Pattern.compile("(\\r\\n|\\r|\\n)");
    private static final String[] COMMON_INDENTS;
    static {
        val count = 32;
        val sb = new StringBuilder(count);
        COMMON_INDENTS = new String[count];
        for (int i = 0; i < count; i++) {
            COMMON_INDENTS[i] = sb.toString();
            sb.append(" ");
        }
    }

    public static CharSequence prefixWithTextBlockEscape(int indent, CharSequence marker, CharSequence content, boolean indentFirst, boolean prefixFirst) {
        val indentStr = indent >= 0 ? indent < COMMON_INDENTS.length ? COMMON_INDENTS[indent] : " ".repeat(indent) : "";
        val parts = Arrays.asList(NL_MATCHER.split(content, -1));
        val result = new StringBuilder(content.length() + marker.length() * parts.size() + indentStr.length() * parts.size());
        if (indentFirst) {
            result.append(indentStr);
        }
        if (prefixFirst) {
            result.append(marker);
        }
        result.append(parts.getFirst());
        for (val part: parts.subList(1, parts.size())) {
            result.append("\n").append(indentStr).append(marker).append(part);
        }
        return result;
    }

    @UtilityClass
    private static class Escaper {
        private static final Int2IntMap ESC_TO_CODE = new Int2IntOpenHashMap();
        static {
            ESC_TO_CODE.put('n', '\n');
            ESC_TO_CODE.put('r', '\r');
            ESC_TO_CODE.put('t', '\t');
            ESC_TO_CODE.put('\\', '\\');
            ESC_TO_CODE.put('"', '"');
            ESC_TO_CODE.put('\'', '\'');
        }
        static int findEscapementLength(@NotNull CharSequence text, int pos) {
            if (pos + 1 < text.length() && text.charAt(pos) == '\\') {
                char c = text.charAt(pos + 1);
                return switch (c) {
                    case 'x' -> 3;
                    case 'u' -> {
                        if (pos + 3 >= text.length() || text.charAt(pos + 2) != '{') {
                            throw new IllegalArgumentException("Invalid unicode escape sequence");
                        }
                        int digits = 0;
                        while (pos + 3 + digits < text.length() && text.charAt(pos + 3 + digits) != '}') {
                            digits++;
                        }
                        yield 3 + digits;
                    }
                    default -> 1;
                };
            } else {
                throw new IllegalArgumentException("This is not an escapement start");
            }
        }

        static int toUnicodeChar(@NotNull CharSequence text, int pos, int length) {
            if (length > 1) {
                val s = switch (text.charAt(pos + 1)) {
                    case 'x' -> text.subSequence(pos + 2, Math.min(text.length(), pos + length + 1));
                    case 'u' -> text.subSequence(pos + 3, Math.min(text.length(), pos + length));
                    default -> throw new AssertionError();
                };
                try {
                    return Integer.parseInt(s.toString(), 16);
                } catch (NumberFormatException e) {
                    return 63;
                }
            } else {
                val c = text.charAt(pos + 1);
                return ESC_TO_CODE.getOrDefault(c, c);
            }
        }
    }
}

File: modules/zig/src/main/kotlin/com/falsepattern/zigbrains/zig/util/ZigStringUtil.kt
package com.falsepattern.zigbrains.zig.util

import com.intellij.openapi.util.TextRange
import it.unimi.dsi.fastutil.ints.Int2IntOpenHashMap
import java.util.regex.Pattern
import java.util.stream.Collectors
import kotlin.math.min

fun CharSequence.escape(): CharSequence {
    val sb = StringBuilder(length)
    this.codePoints().forEachOrdered {
        sb.append(when (it) {
            '\n'.code -> "\\n"
            '\r'.code -> "\\r"
            '\t'.code -> "\\t"
            '\\'.code -> "\\\\"
            '"'.code -> "\\\""
            '\''.code, ' '.code, '!'.code,
            in '#'.code..'&'.code,
            in '('.code..'['.code,
            in ']'.code..'~'.code
                -> Character.toString(it)

            else -> "\\u{" + Integer.toHexString(it) + "}"
        })
    }
    return sb
}

fun CharSequence.decodeReplacements(
    isMultiline: Boolean,
    inputLength: Int = this.length
): List<Pair<TextRange, String>> {
    if (isMultiline)
        return emptyList()

    val result = ArrayList<Pair<TextRange, String>>()
    var index = 0
    while (index + 1 < inputLength) {
        if (this[index] != '\\') {
            index++
            continue
        }

        val length = findEscapementLength(inputLength, index)
        val charCode = toUnicodeChar(index, length)
        val range = TextRange(index, min(index + length + 1, inputLength))
        result.add(Pair(range, Character.toString(charCode)))
        index += range.length
    }
    return result
}

fun CharSequence.unescape(
    isMultiline: Boolean,
    inputLength: Int = if (isMultiline) 0 else this.length
): CharSequence =
    if (isMultiline) {
        // toString done here to decouple input from output in case input is mutable
        this.toString()
    } else {
        // the returned StringBuilder is already decoupled so it's fine
        processReplacements(decodeReplacements(false, inputLength), inputLength)
    }

fun CharSequence.prefixWithTextBlockEscape(
    indent: Int,
    marker: CharSequence,
    indentFirst: Boolean,
    prefixFirst: Boolean,
    newLineAfter: Boolean
): CharSequence {
    val indentStr = if (indent >= 0) {
        if (indent < COMMON_INDENT_COUNT)
            COMMON_INDENTS[indent]
        else
            " ".repeat(indent)
    } else {
        ""
    }
    val parts = NL_MATCHER.split(this, -1)
    val result = StringBuilder(length + (indentStr.length + marker.length) * parts.size + 1)
    if (indentFirst) {
        result.append('\n').append(indentStr)
    }
    if (prefixFirst) {
        result.append(marker)
    }
    result.append(parts.first())
    val partsSize = parts.size
    for (i in 1..<partsSize) {
        result.append('\n').append(indentStr).append(marker).append(parts[i])
    }
    if (newLineAfter)
        result.append('\n').append(indentStr).append(marker)
    return result
}

private const val COMMON_INDENT_COUNT = 32
private val COMMON_INDENTS = Array(COMMON_INDENT_COUNT) { i ->
    " ".repeat(i)
}

private val NL_MATCHER = Pattern.compile("(\\r\\n|\\r|\\n)")

private val ESC_TO_CODE = Int2IntOpenHashMap().apply {
    this['n'.code] = '\n'.code
    this['r'.code] = '\r'.code
    this['t'.code] = '\t'.code
    this['\\'.code] = '\\'.code
    this['"'.code] = '"'.code
    this['\''.code] = '\''.code
}


private fun CharSequence.processReplacements(
    replacements: List<Pair<TextRange, String>>,
    inputLength: Int = this.length
): CharSequence {
    val result = StringBuilder()
    var currentOffset = 0
    for ((range, replacement) in replacements) {
        result.append(this.subSequence(currentOffset, range.startOffset))
        result.append(replacement)
        currentOffset = range.endOffset
    }
    result.append(this.subSequence(currentOffset, inputLength))
    return result
}

private fun CharSequence.findEscapementLength(length: Int, pos: Int): Int {
    if (pos + 1 < length && this[pos] == '\\') {
        val c = this[pos + 1]
        return when (c) {
            'x' -> 3
            'u' -> run {
                if (pos + 2 >= length || this[pos + 2] != '{') {
                    return@run -1
                }
                var digits = 0
                while (pos + 3 + digits < length && this[pos + 3 + digits] != '}') {
                    digits++
                }
                return@run 3 + digits
            }

            else -> 1
        }
    } else {
        return -1
    }
}

private fun CharSequence.toUnicodeChar(pos: Int, length: Int) =
    if (length > 1) {
        val start: Int
        val end: Int
        when (this[pos + 1]) {
            'x' -> {
                start = pos + 2
                end = pos + length + 1
            }

            'u' -> {
                start = pos + 3
                end = pos + length
            }

            else -> throw AssertionError()
        }
        try {
            Integer.parseInt(this, start, end, 16)
        } catch (_: NumberFormatException) {
            '?'.code
        }
    } else {
        val c = this[pos + 1].code
        ESC_TO_CODE.getOrDefault(c, c)
    }
